File: src/main/java/fr/raconteur/simpleskinswapper/gui/SkinTypeStore.java
package fr.raconteur.simpleskinswapper.gui;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import fr.raconteur.simpleskinswapper.SimpleSkinSwapper;
import net.fabricmc.loader.api.FabricLoader;

import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

public class SkinTypeStore {

    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    private static final Type MAP_TYPE = new TypeToken<Map<String, String>>() {}.getType();

    private static Path typesFile() {
        return FabricLoader.getInstance().getGameDir().resolve("skins").resolve("types.json");
    }

    private static Map<String, String> load() {
        Path path = typesFile();
        if (!Files.exists(path)) return new HashMap<>();
        try {
            String json = Files.readString(path);
            Map<String, String> map = GSON.fromJson(json, MAP_TYPE);
            return map != null ? map : new HashMap<>();
        } catch (IOException e) {
            SimpleSkinSwapper.LOGGER.warn("Could not read types.json: {}", e.getMessage());
            return new HashMap<>();
        }
    }

    private static void save(Map<String, String> map) {
        Path path = typesFile();
        try {
            Files.createDirectories(path.getParent());
            Files.writeString(path, GSON.toJson(map));
        } catch (IOException e) {
            SimpleSkinSwapper.LOGGER.warn("Could not write types.json: {}", e.getMessage());
        }
    }

    /** Returns the stored type for a skin file, falling back to the auto-detected one. */
    public static SkinType getType(String filename, SkinType detected) {
        String stored = load().get(filename);
        if (stored == null) {
            // First access: persist the detected value
            setType(filename, detected);
            return detected;
        }
        return "slim".equals(stored) ? SkinType.SLIM : SkinType.CLASSIC;
    }

    /** Stores the user-chosen type for a skin file. */
    public static void setType(String filename, SkinType type) {
        Map<String, String> map = load();
        map.put(filename, type.getMojangVariant());
        save(map);
    }

    /** Removes the stored type for a skin file, e.g. when the file is deleted. */
    public static void removeType(String filename) {
        Map<String, String> map = load();
        if (map.remove(filename) != null) {
            save(map);
        }
    }
}

File: src/main/kotlin/fr/raconteur/simpleskinswapper/gui/SkinTypeStore.kt
package fr.raconteur.simpleskinswapper.gui

import fr.raconteur.simpleskinswapper.data.FabricSkinLibraryEnv
import fr.raconteur.simpleskinswapper.data.JsonFileStore
import fr.raconteur.simpleskinswapper.data.SkinLibraryEnv
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer

/** Persists per-skin model types (skins/types.json, filename -> "slim"|"classic"). */
class SkinTypeStore(private val env: SkinLibraryEnv) {

    private val store = JsonFileStore(
        fileLabel = "types.json",
        path = { env.skinsDir().resolve("types.json") },
        serializer = MapSerializer(String.serializer(), String.serializer()),
        fresh = { linkedMapOf() },
    )

    /** Returns the stored type for a skin file, falling back to the auto-detected one. */
    fun getType(filename: String, detected: SkinType): SkinType {
        val stored = store.load()[filename]
        if (stored == null) {
            // First access: persist the detected value
            setType(filename, detected)
            return detected
        }
        return if (stored == "slim") SkinType.SLIM else SkinType.CLASSIC
    }

    /** Stores the user-chosen type for a skin file. */
    fun setType(filename: String, type: SkinType) {
        val map = store.load().toMutableMap()
        map[filename] = type.mojangVariant
        store.save(map)
    }

    /** Removes the stored type for a skin file, e.g. when the file is deleted. */
    fun removeType(filename: String) {
        val map = store.load().toMutableMap()
        if (map.remove(filename) != null) {
            store.save(map)
        }
    }

    /** Moves a stored type from the old file name to the new one after a file rename. */
    fun renameType(oldFilename: String, newFilename: String) {
        if (oldFilename == newFilename) return
        val map = store.load().toMutableMap()
        val value = map.remove(oldFilename) ?: return
        map[newFilename] = value
        store.save(map)
    }
}

/** GUI-facing singleton delegating to the production-backed instance. */
object SkinTypes {

    private val instance by lazy { SkinTypeStore(FabricSkinLibraryEnv) }

    @JvmStatic fun getType(filename: String, detected: SkinType): SkinType = instance.getType(filename, detected)
    @JvmStatic fun setType(filename: String, type: SkinType) = instance.setType(filename, type)
    @JvmStatic fun removeType(filename: String) = instance.removeType(filename)
    @JvmStatic fun renameType(oldFilename: String, newFilename: String) = instance.renameType(oldFilename, newFilename)
}
